### ISSUE TITLE
WebView2 blocks mouse input to the application after typing too fast

### ISSUE THREAD

**Issue by ghord** [microsoft/WindowsAppSDK#3266]
### Describe the bug

WebView2 freezes mouse input when typing too fast.

### Steps to reproduce the bug

1. Create new application with WebView2 control
2. Set source to any page with input (google/bing)
3. Type in input very fast
4. Mouse input to whole app freezes (notice unchanging cursors). Only keyboard input continues to work.

### Expected behavior

WebView2 continues to work normally

### Screenshots

![Animation](https://user-images.githubusercontent.com/10825459/209000716-f2a45fc8-06c0-4c0d-a985-87691746e8d2.gif)

### NuGet package version

Windows App SDK 1.2.1: 1.2.221116.1

### Packaging type

Unpackaged

### Windows version

Windows 11 version 22H2 (22621, 2022 Update)

### IDE

Visual Studio 2022

### Additional context

This **does** reproduce on both packaged and unpackaged apps.

This **does not** reproduce on versions 1.1.x. 

This issue occurs in every 1.2.x version I've tried.

**Comment by applefanbois** [comment]
We have this exact problem

**Comment by Eilon** [comment]
To repro this in a trivial WinUI3 app:

1. Create a new C# WinUI3 project in VS
2. Change MainWindow.xaml to have just a WebView2:
    ```xaml
    <WebView2 x:Name="wv2" ></WebView2>
    ```
3. Change MainWindow.xaml.cs to have this code:
    ```c#
        public MainWindow()
        {
            this.InitializeComponent();

            wv2.CoreWebView2Initialized += Wv2_CoreWebView2Initialized;

            wv2.EnsureCoreWebView2Async();
        }

        private void Wv2_CoreWebView2Initialized(WebView2 sender, CoreWebView2InitializedEventArgs args)
        {
            wv2.NavigateToString(@"
    <html>
    <body>
    <h1>Hello, world!</h1>

    <div class=""container"">
        <div class=""row"">
            <div class=""col"">
                <input class=""form-control"" type=""text"" />
            </div>
        </div>
        <div class=""row"">
            <div class=""col"">
                <input class=""form-control"" type=""text"" />
            </div>
        </div>
    </div>
    </body>
    </html>
    ");
        }
    ```
4. Run the app and start typing in the two textboxes while using the mouse to click back and forth between them

Eventually you'll still be able to type, but clicking with the mouse won't work.

Can't really tell from the screenshot, but I can assure you that keyboard/typing works, yet I can't click on the textboxes (it ignores the clicks):

<img width="480" alt="image" src="https://user-images.githubusercontent.com/202643/210863358-420210d7-03fb-4a8c-8862-b2ae74ec313c.png">

**Comment by baaaaif** [comment]
we're affected too
--> upvote...

**Comment by applefanbois** [comment]
Do they even test it more than 2 seconds before releasing it?

To be fair, crApple is not better, last year the WKWebview was crashing all the times and it took 8 months before they fix it. This year Apple WKWebview on macOS Intel will never trigger the onKey events, so every website on the planet with a text input will not work.

It just proves that big corps don't test before releasing and not only the programmers suffers, but in the end their customers suffer from their lack of respect.

**Comment by AdriaanLarcai** [comment]
We are also affected by this problem and our implementation of the Monaco editor. We cannot go live with this issue. Please resolve this ASAP.

**Comment by lgBlog** [comment]
What's going on now?

**Comment by lgBlog** [comment]
Is there any progress on the problem?

**Comment by Xyncgas** [comment]
This would've made bill gates poor back in the days

**Comment by Xyncgas** [comment]
>Bill was a true programmer

Sorry I didn't know better, thanks for the information given

**Comment by codendone** [comment]
Tracking with [internal bug](http://task.ms/43203700).

**Comment by Xyncgas** [comment]
I suspect it's because they are using Task based mechanism to handle user input then send it to the browser like a proxy, and they use some kinds of mutex or lock that doesn't get released properly

**Comment by applefanbois** [comment]
The design of the MS framework is baffling and insane. Everything is async so everything becomes 400x slower than anything on Android and Apple because of the useless context switching (only I should decide what is async). And the biggest sin of the computer world : MS are a huge fan of thread re-entrance where our code running on the main thread will be pause execution mid-way (by MS) and another main thread task from MS will trigger again the same code and complete and then the code that was running will resume.

Instead of putting the next task on the main thread queue, they pause our main. This is criminally insane and dangerous. Another problem is that every MS function can throw and it is never mentioned in the documentation. Worst most of the time you will get unhandled exception inside a try catch block.

Not only it is painful to me an my customers because it require 100x more work to make the app just not crash every 2 minutes, but everything made by MS will be unstable, including WebView2 that crash every seconds.

There are so many wait, dispatch and useless validation just so the app won't crash every second. Anytime you call a MS function, you risk crashing the app. But getting so many unhandled exception inside try catch block is maddening.

Stop throwing MS, just make your code work. There is even thread re-entrance inside the MS library. Found that if you call the HttpClient thousands of time over 1 hour, eventually, deep inside MS code, there will be thread-re-entrance where the same line of code is executed twice and that will crash the app.

**Comment by Xyncgas** [comment]
How come cloud gaming doesn't have this problem

**Comment by Xyncgas** [comment]
20 mins to update, 2 months to fix bug, microsoft is terrible

**Comment by KieranDevvs** [comment]
> Tracking with [internal bug](http://task.ms/43203700).

Internal tracking is great... but can the users please get an update?

**Comment by alelom** [comment]
Just reporting another instance of this bug that is completely deal-breaking for my application developed in MAUI Hybrid.

**Comment by ctigrisht** [comment]
We can't develop our app and was forced to put the windows version of the MAUI blazor project on hold, forced to use it on android device only.

**Comment by Xyncgas** [comment]
> We can't develop our app and was forced to put the windows version of the MAUI blazor project on hold, forced to use it on android device only.

.NET 6 still works

**Comment by ctigrisht** [comment]
> You can update client on windows once it's fixed

It says "fixed internally", does that mean a public fix is coming soon? thanks?

**Comment by KieranDevvs** [comment]
> > We can't develop our app and was forced to put the windows version of the MAUI blazor project on hold, forced to use it on android device only.
> 
> .NET 6 still works On another environment, copy your project and change it all from .NET 7 to .NET 6, then wait for the problem fixed. You can update client on windows once it's fixed

Its not that easy, there are a lot of features and API's that were introduced in .NET 7 that people are using. It turns into a migration project.

**Comment by Xyncgas** [comment]
This bug actually seems very interesting, can you guys tell a story on how it was fixed, what went wrong was it hard

**Comment by applefanbois** [comment]
I bet that it was threading problem. Same code executing again while the previous one has not completed. They should use more TryEnqueue() and put the next job on the thread queue so it will be executed when the thread is free.

**Comment by KieranDevvs** [comment]
> This bug actually seems very interesting, can you guys tell a story on how it was fixed, what went wrong was it hard

Lets just focus on actually getting the bug fix first. 

@gabbybilka What's the point of having an open source community with issue tracking if you're just going to track it internally and never respond to the public?

**Comment by applefanbois** [comment]
Open Source does not always mean the Richard Stallman (St-iGNUsius) definition of  open source.
Anything can be open source with a  good de-assembler like "Hopper Disassembler" I guess.

Open Source is just a generic good sounding name like democracy and peace, but in practice it means quiet the opposite.

**Comment by Xyncgas** [comment]
> Open Source does not always mean the Richard Stallman (St-iGNUsius) definition of open source

Richard said was free is not open source, former is better

**Comment by codendone** [comment]
> This bug actually seems very interesting, can you guys tell a story on how it was fixed, what went wrong was it hard

This was a multi-threaded multi-HWND issue. Mouse/pointer input is delivered to a background thread (to enable things like smooth touch panning of ScrollViewers) while keyboard input still goes to the UI thread. These need to be synchronized. WebView2 introduces an extra HWND for keyboard input, and this caused problems in the synchronization code.

**Comment by Xyncgas** [comment]
> This was a multi-threaded multi-HWND issue. Mouse/pointer input is delivered to a background thread (to enable things like smooth touch panning of ScrollViewers) while keyboard input still goes to the UI thread. These need to be synchronized. WebView2 introduces an extra HWND for keyboard input, and this caused problems in the synchronization code.

That's amazing. Thanks for fixing this bug.

**Comment by true-perfect-code** [comment]
Then through which way the update comes to us developers (Windows, Visual Studio, NET MAUI...)?

Thanks

**Comment by ManbirSinghRakhra** [comment]
When will you release fix?

**Comment by rusfield** [comment]
> Then through which way the update comes to us developers (Windows, Visual Studio, NET MAUI...)?
> 
> Thanks

I was about to ask this question here too, for my MAUI Blazor app. I did some research yesterday, and if I understood it right, WindowsAppSDK is a NuGet we can update whenever the fix is released (or even prereleased).

WindowsAppSDK is installed as a dependency through Microsoft.Maui.Graphics.Win2D.WinUI.Desktop for me, and I see that when I search for WindowsAppSDK in the NuGet search, I am unable to install any of the other version than the one I currently have (they all say they are locked by project). So I asked in one of the MAUI Discords, and someone from Microsoft told me to reference the custom version through the .csproj file. If I understood him right, this is what he meant:
`<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.2.230217.4" Condition="'$(TargetFramework)' == 'net7.0-windows10.0.20348'" />`

I guess the status in this issue will change whenever it is available, so I will do the full research then. But someone feel free to correct me :)

**Comment by true-perfect-code** [comment]
@codendone 
Any news about an update possibility?

**Comment by applefanbois** [comment]
no

**Comment by codendone** [comment]
I don't yet have a release timeline to share beyond we'd like it to be soon. The fix is currently going through extra testing.

**Comment by true-perfect-code** [comment]
Will the update be available to us when a new nuget version of Microsoft.Web.WebView2 is released here?

See https://www.nuget.org/packages/Microsoft.Web.WebView2

**Comment by Ewerton** [comment]
By the way, is there a way of force my .Net Maui App to use a specific version of Microsoft.Web.WebView2?

**Comment by vallgrenerik** [comment]
@Eilon Do you know how we (.NET MAUI Blazor developers) will get this fix? 
Is it by nuget, VS, new SDK? 😊
Thanks in advance and have a nice weekend!

**Comment by codendone** [comment]
The fix will be delivered as a Windows App SDK update.

**Comment by vallgrenerik** [comment]
> The fix will be delivered as a Windows App SDK update.

Hi!
I don't think I have installed any Windows sdk explicitly. I have just installed the Visual studio MAUI workload. Will I get it in a VS update then? 😊

**Comment by Xyncgas** [comment]
> > The fix will be delivered as a Windows App SDK update.
> 
> Hi! I don't think I have installed any Windows sdk explicitly. I have just installed the Visual studio MAUI workload. Will I get it in a VS update then? 😊

yes, Visual Studio is microsoft's powerful IDE for making all kinds of program. When you install a 'workload' you installed what you needed to develop for the application. A couple years / months ago there is Xamarin, their selling point was that you can write UI that run on Android + IOS, while still having access to native UI you are mostly writing 'common' UI features for both platform together. Xamarin achieved this by using  native dependency for each platforms, MAUI too is using WindowsSDK when your app is going to be running on WIndows. When you are updating Visual Studio you are getting the newest version of the SDK for your development.

**Comment by true-perfect-code** [comment]
If I understand you correctly @Xyncgas , we don't have to look for new Nuget versions or donwload SDK (see https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/downloads), we just wait for the next Visual Studio update.

I need to get another thing off at this point. The fact that this problem has dragged on for so long indicates that MAUI is used almost exclusively with XAML. And if someone then wants to display the web content in their app, then they will have the current problem through WebView2. And from the way it seems, it's really only a few programmers. 
On the other hand, if you use MAUI with Blazor, then there is no workaround and with such errors in the VebView component, the developed app is unusable because Blazor cannot work with MAUI without WebView. Two points worry me a lot here:
1. the heavy dependency on WebView from MAUI Blazer or no alternative.
2. the fact that few people realise the benefits of MAUI Blazor. Here I am thinking especially of XAML vs. HTML/CSS. I have been programming with WPF for more than 10 years and I used to work on one or another web project in between. I was very happy a few years ago when Blazor came out, because it has a lot of advantages compared to XAML. Unfortunately, I have the impression that most programmers don't see it that way. Why is that?

Thanks

**Comment by Xyncgas** [comment]
I don't like XAML, it's a static description, where as html you can change it dynamically and add event listeners to elements.
The overhead of parsing plaintext html might seem stupid, it is actually very fast to parse string due to SIMD optimizations.

Everyone seriously wanting a MAUI.Blazor app can switch back to .NET 6 and this problem will not exist, publishers can choose the publish a webview that's evergreen or with the version locked.

> the heavy dependency on WebView from MAUI Blazer or no alternative.

The dependency for MAUI.Blazor to have webview is so there is a browser to render web content

**Comment by koenvd** [comment]
Tried to get some feedback in the WinUI community call but had no luck @gabbybilka  (I noticed as well the 1.3 milestone was added).  But any update when a fix would become available? cc @codendone

**Comment by gabbybilka** [comment]
Fixed and available in 1.2.5: https://www.nuget.org/packages/Microsoft.WindowsAppSDK/1.2.230313.1 Will be available in 1.3.0 when it ships as well.

**Comment by AdriaanLarcai** [comment]
> Fixed and available in 1.2.5: https://www.nuget.org/packages/Microsoft.WindowsAppSDK/1.2.230313.1 Will be available in 1.3.0 when it ships as well.

Thank you!

**Comment by koenvd** [comment]
Awesome! Thanks!!

**Comment by KieranDevvs** [comment]
@Eilon can we get `Microsoft.WindowsAppSDK` upgraded to this version ([1.2.230313.1](https://www.nuget.org/packages/Microsoft.WindowsAppSDK/1.2.230313.1)) on MAUI?

**Comment by DierkDroth** [comment]
Alright! At first glance a long standing issue which as road block for me is resolved now https://github.com/MicrosoftEdge/WebView2Feedback/issues/3167

Fingers crossed!

**Comment by ManbirSinghRakhra** [comment]
How do we update it in MAUI. I have seen the nuget package manager is saying this package is blocked by the project.
I have tried to add it in the csproj manually. 

Do we need to do any thing special for updating it inside Maui Blazor App.

**Comment by rusfield** [comment]
> How do we update it in MAUI? I have seen the nuget package manager is saying this package is blocked by the project. I have tried to add it in the csproj manually.
> 
> Do we need to do any thing special for updating it inside Maui Blazor App?

In my MAUI Blazor csproj, I have added this package reference manually into an <ItemGroup>, and it seems to have done it for me (at least no freezes so far):
`<PackageReference Include="Microsoft.WindowsAppSDK" Version="1.2.230313.1" />`

I was also struggling with this "blocked by project" thing and could not do it through the package manager.
I think you can verify whether the new version gets installed by expanding your Blazor MAUI project > Dependencies > Packages > Microsoft.WindowsAppSDK (version), in Visual Studio after rebuilding, to verify that its updated.

**Comment by true-perfect-code** [comment]
I have updated my Visual Studio to version 17.5.3 and the error is fixed on Windows 10.

Unfortunately, it did not work on Windows Server 2019. I have written a new post about this here (but no feedback from WebView2 team yet). See https://github.com/MicrosoftEdge/WebView2Feedback/issues/3333

**Comment by Xyncgas** [comment]
If I create a new MAUI.Blazor project, does this problem still exist

**Comment by lvendramini** [comment]
> If I create a new MAUI.Blazor project, does this problem still exist
yes

**Comment by Eilon** [comment]
See this comment for how to get the fix in a .NET MAUI Blazor project: https://github.com/MicrosoftEdge/WebView2Feedback/issues/3003#issuecomment-1499349079

**Comment by Xyncgas** [comment]
is this included inside 1.3.230502000
